docs/pii-encryption.md

```diff
@@ -0,0 +1,126 @@
+a server-controlled random string, and a hardware security module (HSM)
+to create an encryption key.
+Any time the user gives us sensitive personal information
+we use that encryption key to encrypt the personal information.
+The sensitive PII may only be decrypted if all three factors are present: the HSM,
+the user's password, and the random string.
+
+Based on consultation with the [National Institute of Standards and Technology](https://www.nist.gov/)
+we follow these steps to create the encryption key:
+
+* generate and store a random, 160-bit string as `salt`
+* using the password and the `salt`, create a [SCrypt](https://github.com/pbhogan/scrypt) hash
+* split the SCrypt hash into two 128-bit strings, `Z1` and `Z2`
+* generate a random 256-bit string `AssignedSecret` (`R`)
+* using the HSM, encrypt `AssignedSecret` as `EncryptedAssignedSecret` (`encrypted_R`)
+* calculate the [XOR](https://en.wikipedia.org/wiki/Exclusive_or)
+  of `EncryptedAssignedSecret` and `Z1` as `MaskedCiphertext` (`D`)
+* store `MaskedCiphertext` with the user record
+* create a SHA-256 hash of the concatenation of `Z2` and `MaskedCiphertext` called `CEK` (`E`)
+* create a SHA-256 hash of `CEK` called `PasswordHash`
+* store `PasswordHash` with the user record
+* using the key `CEK`, encrypt the PII using AES GCM 256-bit encryption
+  as `EncryptedPII` (`C`)
+* store `EncryptedPII` with the user record
+* do *not* store plaintext PII, `AssignedSecret` or `CEK`
+
+You can review [the tests for this model in our public repository](../spec/services/pii/nist_encryption_spec.rb).
+
+Since the user's password is an integral part of this multi-factor
+model, if the user forgets their password, the PII may not be recovered.
+To mitigate that loss, we repeat the process above using the user recovery
+code, a randomly generated 128-bit string the user is given when they create
+their account. The recovery code can be used once as a replacement for their
+second authentication factor (e.g. if they lose their mobile phone). It also functions
+as a backup password for encrypted PII.
+
+On login, the decryption process is applied, and the PII is re-encrypted
+using a server-controlled key and placed in the session to permit further
+access without requiring the password. The server-controlled key is a symmetric
+key which is rotated regularly.
+
+When PII needs to be accessed (e.g. to show a user their SSN, or to assert to a
+government relying party) the ciphertext is
+read from the session, decrypted using the server-controlled key, and the
+particular data elements needed are extracted.
+
+When sessions expire the ciphertext they contain is deleted.
+
+When a user loses their password and requests a reset, if the user has
+encrypted sensitive PII stored with us, they are also prompted for their recovery code.
+The recovery code acts as a backup password. Their PII can be decrypted,
+a new recovery code generated, and the PII re-encrypted.
```

Comment:
this is a good start but I think the explanation could be a bit more verbose, including what happens if the user does not have the recovery code.